### PR TITLE
chore(hybrid-cloud): test model decorators and fk breakage

### PR DESCRIPTION
**Pull request by corps**
Validates that all registered models have silo decorators, and that no foreign keys span silos.

It has a helpful exception override which is used to handle our current unresolved work.

Main goal here is to document which key spans need breaking, and also to prevent future ones from being created.

### CODE REVIEW

**Comment by corps** [comment]
Follow up to get this into getsentry, but much smaller changes.

**Comment by corps** [comment]
Test failures are related to model decorator changes that go from control_silo_with_replication -> only.

File: src/sentry/sentry_metrics/indexer/postgres/models.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 from typing import Mapping, Type
 
 
-@control_silo_with_replication_model
+@control_silo_only_model
```
Comment (markstory) [inline src/sentry/sentry_metrics/indexer/postgres/models.py:18]: This should be a region model as metrics indexer is part of ingestion.

File: tests/sentry/db/test_silo_models.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+    def __ge__(self, other):
+        return False
+
+    def __hash__(self):
+        return 1
+
+
+decorator_exemptions = {}
+fk_emeptions = {
```
Comment (markstory) [inline tests/sentry/db/test_silo_models.py:83]:
```suggestion
fk_exemptions = {
```

File: tests/sentry/db/test_silo_models.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    (User, ExportedData),
+}
+
+
+def test_models_have_silos():
+    validate_models_have_silos(decorator_exemptions)
+
+
+def test_silo_foreign_keys():
+    validate_no_cross_silo_foreign_keys(fk_emeptions)
```
Comment (markstory) [inline tests/sentry/db/test_silo_models.py:186]:
```suggestion
    validate_no_cross_silo_foreign_keys(fk_exemptions)
```

File: tests/sentry/hybrid_cloud/__init__.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+from typing import Set, Tuple, Type
+
+from django.db.models import Model
+from django.db.models.fields.related import RelatedField
+
+from sentry.db.models.base import ModelSiloLimit
+from sentry.silo import SiloMode
+
+
+def _iter_mdoels():
```
Comment (markstory) [inline tests/sentry/hybrid_cloud/__init__.py:10]:
```suggestion
def _iter_models():
```

File: tests/sentry/hybrid_cloud/__init__.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    for app, app_models in apps.all_models.items():
+        for model in app_models.values():
+            if model.__module__.startswith("django."):
+                continue
+            yield model
+
+
+def validate_models_have_silos(exemptions: Set[Type[Model]]):
+    for model in _iter_mdoels():
```
Comment (markstory) [inline tests/sentry/hybrid_cloud/__init__.py:21]:
```suggestion
    for model in _iter_models():
```

File: tests/sentry/hybrid_cloud/__init__.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            SiloMode.REGION not in model._meta.silo_limit.modes
+            and SiloMode.CONTROL not in model._meta.silo_limit.modes
+        ):
+            raise ValueError(
+                f"{model!r} is marked as a pending model, but either needs a placement or an exemption in this test."
+            )
+
+
+def validate_no_cross_silo_foreign_keys(exemptions: Set[Tuple[Type[Model], Type[Model]]]):
+    for model in _iter_mdoels():
```
Comment (markstory) [inline tests/sentry/hybrid_cloud/__init__.py:38]:
```suggestion
    for model in _iter_models():
```

**Review by markstory — APPROVED** [review]
Looks good to me.

**Pull request merged**